### ISSUE TITLE
fcl::Halfspace is bad as a dynamic geometry

### ISSUE THREAD

**Issue by SeanCurtis-TRI** [RobotLocomotion/drake#19263]
### What happened?

This ultimately comes from an internal slack conversation.

tl;dr Using `fcl::Halfspace` as a *dynamic* geometry for `SceneGraph` can lead to segfaults.

## User-facing Details

The user defined an sdf file (abbreviated below):
```xml
<sdf version="1.7">
  <model name="ground_plane">
    <static>1</static>
    <link name="ground">
      <collision name="collision">
        <geometry>
          <plane><normal>0 0 1</normal><size>100 100</size></plane>
        </geometry>
      </collision>
    </link>
  </model>
</sdf>
```

When this gets added to `MultibodyPlant`, the `<static>1</static>` makes it so the `ground` body gets welded to the world body. 

### FCL gets angry

In `fcl`, each geometry declares a "local" AABB. In the case of the `fcl::Halfspace` the extents of the bounding box go out to `std::numeric_limits<S>::max()` where the half space has infinite reach. There are several operations on this bounding box that can easily lead to a bounding box with *infinite* values (e.g., translating the bounding box, rotating the bounding box with rounding error, etc.)  Furthermore, there are further `fcl` operations on bounding boxes that aren't compatible with this maximal representation -- `fcl` will compute the radius of a circumscribing sphere of the box.

The dynamic geometries are included in a hierarchy that gets updated many times. The process of updating can lead to the massive (but finite) `fcl::Halfspace` bounding box being promoted to becoming infinite. This infinite bounding box can interfere with the hierarchy operations and, in the worst case, lead to a segfault.

We want to make sure we don't trip over this by passing geometries with infinite bounding boxes to `fcl`.

### How did the `fcl::Halfspace` end up as a *dynamic* geometry

The user had introduced a geometry that ultimately got welded to the world. Surely that means its an anchored geometry (that is, after all, what "anchored" means)? 

During parsing, the geometry is registered as *dynamic* geometry (via `MbP`). At the time the geometry is parsed, there is no context knowing that it is *going* to get anchored. During `MbP::Finalize()` `MbP` examines the graph and finds bodies that are welded (directly or indirectly) to the world. The only thing it does with this information is introduce *collision filters*. Geometries otherwise registered as *dynamic* are not down converted to be *anchored*.

## Resolution

A collection of things that can/should be done to improve the world. We don't need to do *all* of them to at least prevent segfaults.

1. When `MbP` recognizes geometry has become anchored, it shouldn't just filter collisions with anchored geometries, it should move dynamic to anchored geometries. Currently, there is no `SG` API that would allow this act. Furthermore, it would have to be done carefully in making sure that contact is still reported in terms of the bodies that have actually be registered.
2. `fcl::Halfspace` is generally not articulated in a way that is generally nice for the dynamic BVH. It would simply be better if Drake simply didn't introduce it. To that end, we could/should represent half spaces as *really big boxes* (to `fcl`). Even if we created a box that was multiple kilometers across, that would be sufficiently infinite w.r.t. robotics applications and still be nowhere near getting close to saturating floating point values.
3. The `fcl` code that maybe dies on infinite bounding boxes should be updated:
    - Minimally, perform a test to make sure we don't segfault and throw an exception instead.
    - Alternatively, reframe it so that bounding boxes with infinite extents don't cause it to die.

**Comment by jwnimmer-tri** [comment]
Segfault traps are no fun.  We should at minimum turn the segfault into an exception.

**Comment by RussTedrake** [comment]
Is there any code path for MbP that registers anchored geometry with SceneGraph?  I hit this, even though my HalfSpace was registered (via RegisterCollisionGeometry) with the world_body().

**Comment by amcastro-tri** [comment]
Ah..., correct. We actually have a (very old!) [TODO about this](https://github.com/RobotLocomotion/drake/blob/b95c19c62d2eab2c2a6a3e37d11788e696b2111f/multibody/plant/multibody_plant.cc#L1011-L1013) @RussTedrake. 

Probably now that our models are only getting bigger and bigger, this is something we should resolve. A posibility could be to have a "lazy" registration where MbP waits until "Finalize()" (when we already have the multibody graph) to call the corresponding SceneGraph APIs. 

Half spaces are very common in robotics, we should support them.

**Comment by jwnimmer-tri** [comment]
> A possibility could be to have a "lazy" registration where MbP waits until "Finalize()" (when we already have the multibody graph) to call the corresponding SceneGraph APIs.

No.  This would break tons of workflows.

Sean listed the steps to resolving this in the OP.  Those are the correct steps to follow, probably in reverse order -- first trap the segfault, then change proximity engine to register `geometry::HalfSpace` with FCL using `fcl::Box` instead of a `fcl::HalfSpace` when when its a dynamic geometry, and then (later) work to change welded geometries to be static.

**Comment by SeanCurtis-TRI** [comment]
FTR Option (2) would be the fastest to resolve, from a coding perspective (thirty minutes maybe).
